src/signature.c

```diff
@@ -1330,16 +1349,73 @@ static struct {
     { CK_UNAVAILABLE_INFORMATION, 0, NULL },
 };
 
+static unsigned char *
+p11prov_rsapss_encode_algorithm_id(P11PROV_SIG_CTX *sigctx, size_t *outlen)
+{
+    RSA_PSS_PARAMS *pss = NULL;
+    unsigned char *buffer = NULL, *p;
+    const P11PROV_MECH *mech;
+    const EVP_MD *md, *mgf1_md;
+    int ret, len;
+
+    /* when OpenSSL calls this before it makes the signature, we still might not
+     * have the defaults set so do it now */
+    ret = pss_defaults(sigctx, NULL, false);
+    if (ret != CKR_OK) {
+        P11PROV_raise(sigctx->provctx, ret, "Failed to set PSS defaults");
+        goto err;
+    }
+
+    ret = p11prov_mech_by_mechanism(sigctx->pss_params.hashAlg, &mech);
+    if (ret != CKR_OK) {
+        P11PROV_raise(sigctx->provctx, ret, "Failed to get mech for digest %lx",
+                      sigctx->pss_params.hashAlg);
+        goto err;
+    }
+    md = mech->hash_md();
```

Comment:
these mds here are never actually used, so we go through several cycles of fetching stuff when all we need is just the digest OID to encode in the ASN.1 structure later.
I wonder if there is a way to just pass the OID or the name directly and avoid a bunch of churn.

Reply:
TBH, unless PSS_PARAMS has any dynamic content, I would seriously considering pre-generating all the options once offline, storing the generated blobs in a static variable.
Then just copying them in place or do a quick DER -> ASN1_OBJECT dance once on first use if ASN1_OBJECT is all that OpenSSL can accept to add this data to the cert.

Reply:
got rid of the EVP_MD.